fetch.bs

```diff
@@ -8449,6 +8597,60 @@ with a <var>promise</var>, <var>request</var>, <var>responseObject</var>, and an
 </div>
 
 
+<h3 id=fetch-later-method>FetchLater method</h3>
+
+<pre class=idl>
+
+dictionary DeferredRequestInit : RequestInit {
+  DOMHighResTimeStamp backgroundTimeout;
+};
+
+partial interface mixin WindowOrWorkerGlobalScope {
+  [NewObject] Promise&lt;Response> fetchLater(RequestInfo input, optional DeferredRequestInit init = {});
```

Comment:
I think the Promise model has an unavoidable race caused by getting JS involved in delivering this state change information.

Here's the most direct way I can think of to trigger the issue

```js
let completed = false;

fetchLater(...).then(() => {completed = true});

addEventListener("pagehide", () => {
  // This task will run first-thing after `pageshow`
  setTimeout(() => {
    if (!completed) {
      doSomething();
    } 
  }, 0);
});
```

Now let's say the fetch occurs while in BFCache. The first task to run on restore from BFCache will see complete as still `false` and will call `doSomething` incorrectly.

Even if we somehow make the tasks that run the promise resolves special so that they run first, you can recreate the same problem using 2 `fetchLater`s. One of the resolve callbacks must run first and would see the wrong value for the `completed` of the other fetch.

It might also be that any `pageshow` event handler will run before these callbacks. I can't find a clear answer to that. Document reactivation is [here](https://html.spec.whatwg.org/multipage/browsing-the-web.html#document-state:reactivate-a-document) but I'm not sure where the task queue is unfrozen.

This race was avoided in the original design because `isPending()` had access to the true state. It was all handled in native code which is allowed to execute while in BFCache - the timer expires, the renderer tells the browser to send and updates the pending state.

I think the only way to avoid this issue is to have an object that gives access to the state and update that object outside of JS.

Reply:
The task that resolves the promise (call `fetch group`'s activate) should be called when reactivating the document, before suspended timers are resumed and before `pageshow`. Both of these are done in specified location so there shouldn't be a race. The race here is perhaps created because of the Userland JS implemetation of pending beacon, it can be solved in user-land in the same way that this kind of race can be avoided with regular fetches.

Reply:
Saying that we will run these tasks before everything else is concerning. Up until now, no JS runs between `pagehide` and `visibilitychanged` (surprisigly`visibilitychanged` fires before `pagehide` *and* before `pageshow`). Giving these promise resolutions special priority doesn't seem right. What if another feature also requires special priority to run before veverything else?

It also doesn't solve the problem when there are 2+ of them. I would hope that people don't put complex code into these resolve callbacks such that it might look at another pending fetch but any amount of work, including updating logged data, can be hidden behind some framework's function call.

Reply:
Hmm I see your point. @annevk how would you feel about having the signature be `DeferredFetchResult fetchLater(...)`, where `DeferredFetchResult` has a `{ sent: false }` that turns to true synchronously at the appropriate moment? (BFCache restore after send / re-show after timers frozen)

Reply:
friendly ping @annevk

Reply:
I'm not sure why this is assuming that the fetch would have happened while in bfcache. If the user goes back quickly that won't be the case, though it might still happen due to it already being enqueued for delivery for instance.

I think you would need a class that exposes both sent state and a promise for when that sent state is ready.

However, if `fetchLater()` doesn't return a promise how do we deal with network errors and policy violations and such?

Reply:
> I'm not sure why this is assuming that the fetch would have happened while in bfcache. If the user goes back quickly that won't be the case, though it might still happen due to it already being enqueued for delivery for instance.

At the very least, it *could* happen while in BFCache.
> 
> I think you would need a class that exposes both sent state and a promise for when that sent state is ready.

That would work!

> 
> However, if `fetchLater()` doesn't return a promise how do we deal with network errors and policy violations and such?

You generally don't; you only care about whether the fetch is transmitted. As in the regular unload case (not bfcache) you don't get this promise.

Reply:
So will change the PR to `{sent: Promise<void>; didSend: boolean }` or so

Reply:
My (current) favorite color:
```webidl
interface FetchLaterState {
  readonly attribute boolean sent;
  readonly attribute Promise<undefined> ready;
}
```

Reply:
> My (current) favorite color:
> 
> ```webidl
> 
> interface FetchLaterState {
> 
>   readonly attribute boolean sent;
> 
>   readonly attribute Promise<undefined> ready;
> 
> }
> 
> ```

Yea I like that

Reply:
In second look I think @fergald is right... A boolean should be enough.
@annevk what is the use case for the promise? Perhaps we can defer that for later and have it be:

```webidl
interface FetchLaterState {
  readonly attribute boolean sent;
}
```

Reply:
After chatting for a while with @noamr I think I understand the idea now. The boolean would reflect whether the browser called fetch and nothing else. So a network error following that is not something you can observe. That works. We should make sure that the boolean flipping happens at a well-defined time.

Reply:
In reality flipping to true means that the browser intends to send this and there is no longer anything you can do to stop it. In a multiprocess implementation the bit should flip *before* sending starts, otherwise there is a race (you could cancel a request that has already started).

`sent` is actually a little misleading since sending may not have actually started or even be possible (e.g. we went offline at just the wrong moment). We could consider `pending` or `cancellable`.

Reply:
I think we should match optional arguments where the boolean is always supposed to default to false.

Reply:
But this should not be optional as `fetchLater()` should always return `FetchLaterState` for user to tell if it is `sent` (or `pending`) later?

> having the signature be DeferredFetchResult fetchLater(...), where DeferredFetchResult has a { sent: false } that turns to true synchronously at the appropriate moment? 

I don't quite get the synchronous part, how could `FetchLaterState` update itself?

Reply:
@annevk 
> I think we should match optional arguments where the boolean is always supposed to default to false.

Are you saying that whatever name we choose it should be something that starts `false` and flips to `true`?

Reply:
@mingyc 
> I don't quite get the synchronous part, how could `FetchLaterState` update itself?

`FetchLaterState` should not update *during* a JS task (unless if has properties that can update in response to some action taken by JS, e.g. if it was `cancellable` then that could change to `false` if you cancel it) and it should not update internally but only become visible later. It should update *between* JS tasks. I'm not sure if synchronous is the right word for that.

Maybe the way to do this is to be specific about the interaction between `sent` and `signal`:
- if `sent` is `false` then aborting `signal` MUST cancel the request and the browser. Cancellation means that that no packets related to that request are/were ever sent.
- if `sent` is `true` then aborting `signal` has no effect.
- if `sent` is `true` then browser has made/is making/will ASAP make an attempt to send the request (this maybe doesn't need to be stated explicitly)

Reply:
Perhaps this can be simplified if setting this state is still done on the document's event loop. Needs a bit of research

Reply:
@noamr 
> Perhaps this can be simplified if setting this state is still done on the document's event loop. Needs a bit of research

I think if you want to phrase it in those terms then you have to say something like
- when the browser wants to send the request due to timeout
  - it pushes a task onto the XXX event loop
  - when the task runs:
    - if the request has not yet been cancelled, `sent` is set to `true` and the request can start
    - if the request has already been cancelled, nothing further happens
- when the browser wants to send the request due to freeing up quota allow another request to be queued
  - we don't worry about cancelled because if it's already cancelled we would not think about it at all
  - `sent` is set to `true` and the can request start
- when the browser wants to send the request due to the document having been destroyed (or a crash)
  - the request can start

XXX event loop must be one which continues processing even while in BFCache.

Reply:
@fergald with regards to naming, yeah that is what I meant.

We should flip `FetchLaterState`'s boolean in at the start of the task that also dispatches `pageshow` in my opinion. The fetch might have already happened, but having a single synchronization point seems good and gives web developers a reliable API. (And doesn't encourage them to start polling.)

Reply:
@annevk my current thinking about it is that when the timeout passes, we post a task to invoke the fetch to the document's event loop, without associating the task with the document. When that task is handled, we only flip the boolean and perform fetch if the document is not fully active.

Reply:
I don't think we have precedent for tasks that are not associated with a document.

Reply:
Hmm you're right, it would also be racy if the `reactivate` task is already queued. So we need an atomic exchange thingy, something like:
Let `sentState` be "deferred", "scheduled", or "sent", initially "deferred".

When deactivating (in document event loop, after `pagehide`):
- Set `sentState` to "scheduled".
- In parallel:
   - Wait `backgroundTimeout` millis
   - Let `currentState` be the result of atomically exchanging `sentState` with "sent"
   - If `currentState` is `scheduled`, then fetch etc.

When reactivating (in document event loop, before `pageshow`):
   - Let `currentState` be the result of atomically exchanging `sentState` with "deferred"
   - Let fetchLater's `sent` be true if `currentState` is "sent", otherwise false.

Reply:
See new revision.

Reply:
@annev
> I think we should match optional arguments where the boolean is always supposed to default to false.

I agree with this as a principle however I think we're talking about a field in a returned struct, not an optional argument.  It's an initial value not a default value. So we don't need to favour `false`.